# **Data Mining I | <span style="color: darkseagreen;">Exploratory Data Analysis and Pre-Processing</span>**

## A Clustering-Based Exploration of WorkplaceAbsence Patterns

*Group 07*

*Group members* : Francisco Gomes (20221810), Margarida Marchão, Pedro Coimbras and Marta Alves.

<a id="1"></a>

# **1. Importing the datasets and libraries <font>**

In [173]:
#imports
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from pylab import rcParams

from utils1 import (
    calculate_percentage,
    analyze_missingness_patterns,
    diagnose_seasons_from_month,
    secret_missing_values,
)

pd.set_option("display.max_columns", None)

In [174]:
#from google.colab import drive
# drive.mount('/content/drive', force_remount=True)
# path = '/content/drive/MyDrive/Colab Notebooks/DM1/projeto/data/'
path2 = '/Users/franciscogomes/Desktop/Faculdade/Master/Data Mining/Project/reasons_for_absenteeism/data/'

<a id="1"></a>

# **2. Preliminary Data Analysis<font>**

In [ ]:
pd.set_option("display.max_columns", None)
abs = pd.read_csv(path2 + 'absenteeism_data.csv',delimiter=';')
abs = abs.set_index("ID")
abs.head()

Reason for absence Month of absence Day of the week Seasons  \
ID                                                                            
11             Unjustified absence             July         Tuesday  Summer   
36                     Unspecified             July         Tuesday     NaN   
3             Medical consultation             July       Wednesday  Summer   
7   Diseases of the eye and adnexa             July        Thursday     NaN   
11            Medical consultation             July        Thursday  Summer   

    Days since previous absence  Transportation expense  \
ID                                                        
11                          0.0                     289   
36                          0.0                     118   
3                           0.0                     179   
7                           0.0                     279   
11                          0.0                     289   

    Distance from Residence to Work  Estimated commute time Service time  \
ID                                                                         
11                               36                      69           13   
36                               13                      26           18   
3                                51                     108           18   
7                                 5                       5           14   
11                               36                      69           13   

    Years until retirement Date of Birth Disciplinary failure  Education  \
ID                                                                         
11                      32    1992-08-15                   No          1   
36                      15    1975-09-02                  Yes          1   
3                       27    1987-04-08                   No          1   
7                       26    1986-07-25                   No          1   
11                      32    1992-08-15                   No          1   

    Number of children Social drinker Social smoker  Number of pets Weight  \
ID                                                                           
11                   2              Y            No               1     90   
36                   1              Y            No               0     98   
3                    0            Yes            No               0     89   
7                    2            Yes           Yes               0     68   
11                   2            Yes            No               1     90   

    Height  Body mass index  Absenteeism time in hours  
ID                                                      
11     172               30                          4  
36     178               31                          0  
3      170               31                          2  
7      168               24                          4  
11     172               30                          2

#### **Feature Groups Overview**

| **Category** | **Feature Name** | **Description / Notes** |
|---------------|------------------|--------------------------|
| **Absence Information** | Reason for absence | Registered reason for each absence |
|  | Month of absence | Month when the absence occurred |
|  | Day of the week | Weekday of the absence (Monday–Friday) |
|  | Seasons | Season in which the absence occurred |
|  | Days since previous absence | Number of days since last absence |
|  | Absenteeism time in hours | Duration of absence (in hours) |
| **Commute & Transportation** | Transportation expense | Monthly commuting expenses (BRL) |
|  | Distance from Residence to Work | Distance between home and workplace |
|  | Estimated commute time | Estimated duration of daily commute |
| **Work-Related Features** | Service time | Years worked at the company |
|  | Years until retirement | Years remaining until retirement eligibility |
|  | Disciplinary failure | Whether absence violated company policy (binary) |
| **Demographics & Personal Info** | Date of Birth → *Age* | Employee’s date of birth (convert to age) |
|  | Education | Education level (1–High School → 4–Master’s/PhD) |
|  | Number of children | Total number of children |
|  | Social smoker | Smokes socially (binary) |
|  | Social drinker | Drinks socially (binary) |
|  | Number of pets | Number of pets owned |
| **Health & Physical Attributes** | Weight | Weight (kg) |
|  | Height | Height (cm) |
|  | Body mass index (BMI) | Calculated from height and weight |


In [202]:
# Display unique values for each column
for col in abs.columns:
    unique_vals = abs[col].unique()
    n_unique = len(unique_vals)
    print(f"\n{col} — {n_unique} unique values")
    print(unique_vals[:30])


Reason for absence — 29 unique values
['Unjustified absence' 'Unspecified' 'Medical consultation'
 'Diseases of the eye and adnexa' 'Patient follow-up'
 'Injury, poisoning and other consequences of external causes'
 'Certain infectious and parasitic diseases' nan
 'Diseases of the digestive system' 'Diseases of the genitourinary system'
 'Factors influencing health status and contact with health services'
 'Diseases of the respiratory system'
 'Diseases of the musculoskeletal system and connective tissue'
 'Dental consultation'
 'Symptoms, signs and abnormal clinical and laboratory findings'
 'Laboratory examination' 'Blood donation'
 'Diseases of the nervous system' 'Physiotherapy'
 'Congenital malformations, deformations and chromosomal abnormalities'
 'Diseases of the ear and mastoid process'
 'Diseases of the skin and subcutaneous tissue'
 'Mental and behavioural disorders' 'Diseases of the circulatory system'
 'Pregnancy, childbirth and the puerperium'
 'Endocrine, nutritional an

In [176]:
display(abs.shape)

(800, 21)

| **Type** | **Count** |
|-----------|-----------|
| Observations (rows) | 800 |
| Variables (columns) | 21 |

Each record corresponds to **one employee’s absence instance**, including personal, work, commute, and health-related information.



<a id="1"></a>

### **2.1. Data Types <font>**

In [177]:
display(abs.info())

<class 'pandas.core.frame.DataFrame'>
Index: 800 entries, 11 to 35
Data columns (total 21 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Reason for absence               740 non-null    object 
 1   Month of absence                 737 non-null    object 
 2   Day of the week                  740 non-null    object 
 3   Seasons                          573 non-null    object 
 4   Days since previous absence      781 non-null    float64
 5   Transportation expense           800 non-null    int64  
 6   Distance from Residence to Work  800 non-null    int64  
 7   Estimated commute time           800 non-null    int64  
 8   Service time                     800 non-null    object 
 9   Years until retirement           800 non-null    int64  
 10  Date of Birth                    800 non-null    object 
 11  Disciplinary failure             800 non-null    object 
 12  Education                  

None

 - missing values in: reason of absence, month of absence, day of the week, seasons, days since previous absence
 - education is int
 - we can predict seasons with month of absence
 - transportation expense is in brasilian reais -> mudar para euros
 - service time and weight has missing values '-' -> mudar para int
 - one hot no disciplinary failure, social frinker and smoker
 - social drinker and smoker has values of 'Y'
 - ver 0s na coluna do absenteeism time in hours -> 0h can mean 30min or it wasnt absent
 - maybe mudar int64 para int32
 - dar drop do days since previous absence porque só tem um unico valor que é 0.0
 - ver outliers
 - compute age
 - há duplicates - 48

<a id="1"></a>

### **2.4. Missing Values Analysis <font>**

In [178]:
abs.isna().sum()

Reason for absence                  60
Month of absence                    63
Day of the week                     60
Seasons                            227
Days since previous absence         19
Transportation expense               0
Distance from Residence to Work      0
Estimated commute time               0
Service time                         0
Years until retirement               0
Date of Birth                        0
Disciplinary failure                 0
Education                            0
Number of children                   0
Social drinker                       0
Social smoker                        0
Number of pets                       0
Weight                               0
Height                               0
Body mass index                      0
Absenteeism time in hours            0
dtype: int64

In [179]:
missing_columns = ['Reason for absence', 'Month of absence', 'Day of the week', 'Seasons', 'Days since previous absence']
missing_percentages = calculate_percentage(abs, columns=missing_columns)

for col, percent in missing_percentages.items():
    display(f'Around {percent}% of the data points in the column {col} are missing values')

'Around 7.5% of the data points in the column Reason for absence are missing values'

'Around 7.9% of the data points in the column Month of absence are missing values'

'Around 7.5% of the data points in the column Day of the week are missing values'

'Around 28.4% of the data points in the column Seasons are missing values'

'Around 2.4% of the data points in the column Days since previous absence are missing values'

estamos a ver o porque de algumas variaves serem object ------------------------------------------------------------------------------

| **Column** | **Missing Values** | **Missing (%)** | **Notes / Observations** |
|-------------|--------------------|-----------------|---------------------------|
| Reason for absence | 60 | 7.5% | Moderate number of missing entries — may relate to data entry issues or unclassified absences |
| Month of absence | 63 | 7.9% | Slightly higher missing ratio; may correspond to missing absence dates |
| Day of the week | 60 | 7.5% | Matches missing patterns in "Reason for absence" — possibly same missing records |
| Seasons | 227 | 28.4% | **Highest missing rate**; may require imputation or removal |
| Days since previous absence | 19 | 2.4% | Minor; mostly zeros, might not significantly impact clustering |
| **All other columns** | 0 | 0% | Fully complete |

In [180]:
cols_absence = ['Reason for absence', 'Month of absence', 'Day of the week', 'Seasons', 'Days since previous absence']
missing_summary = analyze_missingness_patterns(abs, cols_absence)

Pairwise Missingness Correlation Matrix:



 Rows with all of Reason/Month/Day missing: 60

 Top 10 Pairs with Highest Overlapping Missing Values:


count
Month of absence   Seasons                         62
Reason for absence Month of absence                60
                   Day of the week                 60
                   Seasons                         60
Month of absence   Day of the week                 60
Day of the week    Seasons                         60
Seasons            Days since previous absence      4
Reason for absence Days since previous absence      0
Month of absence   Days since previous absence      0
Day of the week    Days since previous absence      0


Example indices where Reason/Month/Day are all missing:


Index([20, 20, 3, 10, 14, 3, 20, 15, 34, 28, 34, 20, 33, 17, 28, 36, 3, 20, 11,
       11],
      dtype='int64', name='ID')

### Missingness Correlation Analysis

| Variable Pair | Correlation | Interpretation |
|----------------|-------------|----------------|
| Reason ↔ Month | 0.97 | Strongly correlated missingness |
| Reason ↔ Day | 1.00 | Identical missing rows |
| Month ↔ Day | 0.97 | Same missing pattern |
| Reason/Month/Day ↔ Seasons | 0.45 | Partial overlap |
| Days since previous absence ↔ others | ≈ 0.0 | No relationship |

The columns `Reason for absence`, `Month of absence`, and `Day of the week` have **perfectly overlapping missing rows (n = 60)**.  
These likely represent **unlogged or incomplete absence records**.  
`Seasons` has additional missing cases, possibly due to inconsistent month data, and `Days since previous absence` appears independent and uninformative.


In [181]:
result_seasons = diagnose_seasons_from_month(abs)

Original missing 'Seasons': 227
Potentially recoverable from 'Month of absence': 165 (72.7%)
Rows where existing 'Seasons' disagrees with derived value: 366


### Diagnostic Results — Seasons Derived from Month of Absence

| **Metric** | **Value** | **Interpretation** |
|-------------|-----------|--------------------|
| Original missing `Seasons` | **227** | About 28.4% of all entries lacked a season label. |
| Potentially recoverable from `Month of absence` | **165 (≈72.7%)** | A large portion of missing `Seasons` can be reliably inferred from the corresponding month using the Brazilian seasonal mapping. |
| Rows where existing `Seasons` disagree with derived value | **366** | There are inconsistencies between the recorded `Seasons` and the season logically expected from the month. |

- Roughly **three-quarters of missing `Seasons` values** could be **reconstructed** based on the month, confirming that the variables are strongly related.  
- However, the presence of **366 mismatched cases** suggests **data quality issues**, likely due to inconsistent labeling or incorrect season assignments.  
- These discrepancies highlight the need for **careful validation** before applying imputation or transformation in the preprocessing phase.  
- At this EDA stage, we only **identify** these inconsistencies — the actual correction will be handled later during **feature engineering**.


In [182]:
cols_with_dash, rows_with_dash, mask_dash = secret_missing_values(abs)
display(f'Columns with dash as missing value: {cols_with_dash}')

Columns containing '-' values:


Weight          77
Service time    28
dtype: int64


Total rows containing '-': 104


Reason for absence Month of absence  \
ID                                                                       
14  Injury, poisoning and other consequences of ex...             July   
3   Factors influencing health status and contact ...             July   
30                                Dental consultation           August   
5   Injury, poisoning and other consequences of ex...        September   
34                               Medical consultation        September   
..                                                ...              ...   
10  Injury, poisoning and other consequences of ex...            April   
28  Injury, poisoning and other consequences of ex...            April   
28                     Diseases of the nervous system              May   
28                                                NaN              NaN   
5                                 Unjustified absence             July   

   Day of the week Seasons  Days since previous absence  \
ID                                                        
14          Monday  Summer                          0.0   
3           Monday  Summer                          0.0   
30       Wednesday  Summer                          0.0   
5          Tuesday     NaN                          0.0   
34          Monday  Summer                          0.0   
..             ...     ...                          ...   
10         Tuesday  Winter                          0.0   
28          Monday     NaN                          0.0   
28       Wednesday  Winter                          0.0   
28             NaN     NaN                          0.0   
5        Wednesday  Summer                          0.0   

    Transportation expense  Distance from Residence to Work  \
ID                                                            
14                     155                               12   
3                      179                               51   
30                     157                               27   
5                      235                               20   
34                     118                               10   
..                     ...                              ...   
10                     361                               52   
28                     225                               26   
28                     225                               26   
28                     225                               26   
5                      235                               20   

    Estimated commute time Service time  Years until retirement Date of Birth  \
ID                                                                              
14                      28           14                      31    1991-09-10   
3                      108           18                      27    1987-04-08   
30                      45            6                      36    1996-05-22   
5                       38            -                      22    1982-06-19   
34                      22           10                      28    1988-05-10   
..                     ...          ...                     ...           ...   
10                     114            3                      37    1997-07-08   
28                      50            -                      37    1997-08-03   
28                      50            -                      37    1997-08-03   
28                      50            -                      37    1997-08-03   
5                       38            -                      22    1982-06-19   

   Disciplinary failure  Education  Number of children Social drinker  \
ID                                                                      
14                   No          1                   2            Yes   
3                    No          1                   0            Yes   
30                   No          1                   0            Yes   
5                    No          1                   1           

'Columns with dash as missing value: Weight          77\nService time    28\ndtype: int64'

### Detecting Hidden Missing Values (`'-'`)

Although missing values were already analyzed, additional inspection revealed that some 
fields contain the symbol **`'-'`**, which is not automatically recognized as a missing value by pandas.  
These entries represent **hidden or non-standard missing data** that must be identified and corrected.

In [183]:
cols_with_dash = ['Weight', 'Service time']
abs[cols_with_dash] = abs[cols_with_dash].apply(lambda col: col.astype(str).str.strip().replace('-', np.nan))

In [184]:
abs.isna().sum()

Reason for absence                  60
Month of absence                    63
Day of the week                     60
Seasons                            227
Days since previous absence         19
Transportation expense               0
Distance from Residence to Work      0
Estimated commute time               0
Service time                        28
Years until retirement               0
Date of Birth                        0
Disciplinary failure                 0
Education                            0
Number of children                   0
Social drinker                       0
Social smoker                        0
Number of pets                       0
Weight                              77
Height                               0
Body mass index                      0
Absenteeism time in hours            0
dtype: int64

To ensure data consistency, all occurrences of the symbol `'-'` were replaced 
with proper `NaN` values using the pandas `.replace()` method.

<a id="1"></a>

### **2.5. Summary Statistics - (numerical data) <font>**

In [185]:
abs.describe().T

count       mean        std    min    25%  \
Days since previous absence      781.0    0.00000   0.000000    0.0    0.0   
Transportation expense           800.0  221.92750  66.778732  118.0  179.0   
Distance from Residence to Work  800.0   29.79875  14.875057    5.0   16.0   
Estimated commute time           800.0   59.34875  31.301067    5.0   31.0   
Years until retirement           800.0   26.74250   9.269498   -1.0   25.0   
Education                        800.0    1.28750   0.667383    1.0    1.0   
Number of children               800.0    1.04375   1.120672    0.0    0.0   
Number of pets                   800.0    0.74500   1.323694    0.0    0.0   
Height                           800.0  172.09375   6.058122  163.0  169.0   
Body mass index                  800.0   26.64125   4.267354   19.0   24.0   
Absenteeism time in hours        800.0    7.08750  13.949124    0.0    2.0   

                                   50%    75%    max  
Days since previous absence        0.0    0.0    0.0  
Transportation expense           225.0  260.0  388.0  
Distance from Residence to Work   26.0   50.0   52.0  
Estimated commute time            52.0   94.0  114.0  
Years until retirement            28.0   33.0   38.0  
Education                          1.0    1.0    4.0  
Number of children                 1.0    2.0    4.0  
Number of pets                     0.0    1.0    8.0  
Height                           170.0  172.0  196.0  
Body mass index                   25.0   31.0   38.0  
Absenteeism time in hours          3.0    8.0  120.0

### Summary of Key Numerical Insights

- **Days since previous absence:** Constant (all 0) → no variance, will likely be removed.  
- **Transportation expense / Commute distance / Commute time:** Show wide variation — important indicators of employee mobility and potential absenteeism factors.  
- **Years until retirement:** Mostly consistent but includes a **negative value** → data quality issue to fix later.  
- **Education:** Low variability (most have level 1) → limited clustering value.  
- **Children & Pets:** Mild variation; may reflect personal/lifestyle differences.  
- **BMI:** Moderate variability (mean ≈ 26.6) — possible link to health-related absences.  
- **Absenteeism time in hours:** Skewed with a few extreme values — key variable to analyze and scale properly.

#### Overall Notes
- Some variables (e.g., **Days since previous absence**, **Height**) provide little information.  
- Commute-related and absenteeism measures are **most informative** for later clustering.  
- **Scaling and cleaning** (especially for `Years until retirement`) will be required before modeling.


#### **Removing Column <font>**

In [186]:
display(abs['Days since previous absence'].unique())

array([ 0., nan])

`Days since previous absence`

From the summary statistics, we observe that:

| Metric | Value |
|---------|--------|
| Count | 781 |
| Mean | 0.00 |
| Standard Deviation | 0.00 |
| Min / Max | 0.0 / 0.0 |
| Quartiles (25%, 50%, 75%) | All 0.0 |

- The variable `Days since previous absence` shows **no variation** across observations — all values are zero.  
- This indicates that, in practice, employees are either **not having consecutive absences** or the feature is **not properly recorded** in the dataset.  
- With such low variance, this variable provides **no discriminative power** for clustering or pattern discovery.


In [187]:
abs = abs.drop(columns=['Days since previous absence'])

<a id="1"></a>

### **2.6. Findings in the Categorical Data <font>**

### Summary of Categorical Features

- **Reason for absence:** 28 categories; most common is *Medical consultation* (149 cases). Indicates diverse reasons for absences, but highly imbalanced distribution.  
- **Month of absence:** 12 unique values; *March* is the most frequent (87). Absences may show seasonal trends.  
- **Day of the week:** 6 unique days; *Monday* dominates (159) — consistent with common absenteeism patterns.  
- **Seasons:** 4 categories; *Spring* most common (155), but ~28% missing values — to be reconstructed later.  
- **Service time:** 19 unique values; *18 years* appears most (150). Suggests a workforce with a broad range of tenure.  
- **Date of Birth:** 36 unique values; likely shared birthdays (e.g., placeholder entries). Will be transformed into **Age**.  
- **Disciplinary failure:** Binary; mostly *No* (94.5%) — very few policy violations.  
- **Social drinker / Social smoker:** Both binary with inconsistent labels (“Yes”/“Y”/“No”). Most are *non-smokers (93%)* and *social drinkers (46%)*.  
- **Weight:** 27 unique values; *89 kg* most frequent (110). Needs type correction (`object` → numeric).

#### Key Takeaways
- Some categorical variables (e.g., **Reason for absence**) have many levels — may need **encoding or grouping** before clustering.  
- Binary variables (**Disciplinary failure**, **Smoker/Drinker**) require **label cleaning** for consistency.  
- **Seasons** and **Date of Birth** need transformation during **feature engineering**.  
- Features like **Day of the week** and **Month of absence** may capture **temporal absenteeism trends**.


In [188]:
abs.describe(include = ['O']).T

count unique                   top freq
Reason for absence     740     28  Medical consultation  149
Month of absence       737     12                 March   87
Day of the week        740      6                Monday  159
Seasons                573      4                Spring  155
Service time           772     18                    18  150
Date of Birth          800     36            1987-04-08  122
Disciplinary failure   800      2                    No  756
Social drinker         800      3                   Yes  367
Social smoker          800      3                    No  744
Weight                 723     26                    89  110

most frequent reason for absence is medical consultation

<a id="1"></a>

### **2.7. Duplicates <font>**

In [196]:
abs[abs.duplicated(keep=False)]

Reason for absence Month of absence Day of the week Seasons  \
ID                                                                    
10       Patient follow-up             July          Friday  Summer   
20                     NaN              NaN             NaN     NaN   
20                     NaN              NaN             NaN     NaN   
3                      NaN              NaN             NaN     NaN   
34    Medical consultation          October         Tuesday  Spring   
34    Medical consultation          October         Tuesday  Spring   
10       Patient follow-up         December       Wednesday     NaN   
28    Medical consultation         December         Tuesday  Spring   
10       Patient follow-up         December       Wednesday     NaN   
28                     NaN              NaN             NaN     NaN   
34           Physiotherapy          January          Monday  Autumn   
34           Physiotherapy          January         Tuesday  Autumn   
34           Physiotherapy          January          Monday  Autumn   
34           Physiotherapy          January         Tuesday  Autumn   
28  Laboratory examination         February        Thursday  Autumn   
28                     NaN              NaN             NaN     NaN   
23       Patient follow-up             June        Thursday  Winter   
3                      NaN              NaN             NaN     NaN   
5     Medical consultation          October          Monday  Spring   
5     Medical consultation          October          Monday  Spring   
3                      NaN              NaN             NaN     NaN   
3            Physiotherapy         February       Wednesday  Autumn   
3            Physiotherapy         February          Friday  Autumn   
3            Physiotherapy         February       Wednesday     NaN   
3            Physiotherapy         February       Wednesday     NaN   
33                     NaN              NaN             NaN     NaN   
24     Dental consultation             July         Tuesday  Summer   
10       Patient follow-up             July          Friday  Summer   
24     Dental consultation             July         Tuesday  Summer   
24     Dental consultation        September         Tuesday  Summer   
24     Dental consultation        September         Tuesday  Summer   
33                     NaN              NaN             NaN     NaN   
28    Medical consultation         December         Tuesday  Spring   
22                     NaN              NaN             NaN     NaN   
3                      NaN              NaN             NaN     NaN   
3            Physiotherapy         February       Wednesday  Autumn   
3                      NaN              NaN             NaN     NaN   
28  Laboratory examination         February        Thursday  Autumn   
3            Physiotherapy         February          Friday  Autumn   
3            Physiotherapy            March       Wednesday  Autumn   
3            Physiotherapy            March       Wednesday  Autumn   
22                     NaN              NaN             NaN     NaN   
3            Physiotherapy            March       Wednesday  Autumn   
22           Physiotherapy            April          Friday  Winter   
22           Physiotherapy            April          Friday  Winter   
22           Physiotherapy            April          Friday  Winter   
15     Dental consultation              May        Thursday  Winter   
22           Physiotherapy              May          Friday  Winter   
15     Dental consultation              May        Thursday  Winter   
22           Physiotherapy              May          Friday  Winter   
22                     NaN              NaN             NaN     NaN   
23       Patient follow-up             June        Thursday  Winter   

    Transportation expense  Distance from Residence to Work  \
ID                                                            
10                     361                     

<a id="1"></a>

# **3. Feature Engineering and Transformation<font>**